File: addons/guide-cpp/src/inputs/guide_input_mouse_axis_2d.cpp
#include "guide_input_mouse_axis_2d.h"
#include "guide_input_state.h"

using namespace godot;

void GUIDEInputMouseAxis2D::_bind_methods() {
    ClassDB::bind_method(D_METHOD("_refresh"), &GUIDEInputMouseAxis2D::_refresh);
}

GUIDEInputMouseAxis2D::GUIDEInputMouseAxis2D() {
}

GUIDEInputMouseAxis2D::~GUIDEInputMouseAxis2D() {
}

bool GUIDEInputMouseAxis2D::_needs_reset() const {
    return true;
}

void GUIDEInputMouseAxis2D::_begin_usage() {
    if (_state) {
        _state->connect("mouse_position_changed", Callable(this, "_refresh"));
        _refresh();
    }
}

void GUIDEInputMouseAxis2D::_end_usage() {
    if (_state) {
        _state->disconnect("mouse_position_changed", Callable(this, "_refresh"));
    }
}

void GUIDEInputMouseAxis2D::_refresh() {
    if (_state) {
        Vector2 delta = _state->get_mouse_delta_since_last_frame();
        _value.x = delta.x;
        _value.y = delta.y;
    }
}

bool GUIDEInputMouseAxis2D::is_same_as(const Ref<GUIDEInput> &other) const {
    return other.is_valid() && other->get_class() == get_class();
}

String GUIDEInputMouseAxis2D::_editor_name() const {
    return "Mouse Axis 2D";
}

String GUIDEInputMouseAxis2D::_editor_description() const {
    return "Relative mouse movement on 2 axes.";
}

int GUIDEInputMouseAxis2D::_native_value_type() const {
    return 2; // AXIS_2D
}

GUIDEInput::DeviceType GUIDEInputMouseAxis2D::_device_type() const {
    return DEVICE_MOUSE;
}

File: addons/GuideCS/Inputs/GUIDEInputMouseAxis2D.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEInputMouseAxis2D : GUIDEInput {
	public override bool _NeedsReset() => true;

	public override void _BeginUsage() {
		_State.MousePositionChanged += _Refresh;
		_Refresh();
	}

	public override void _EndUsage() {
		_State.MousePositionChanged -= _Refresh;
	}

	private void _Refresh() {
		Vector2 delta = _State.GetMouseDeltaSinceLastFrame();
		_Value = new Vector3(delta.X, delta.Y, 0);
	}

	public override bool IsSameAs(GUIDEInput other) {
		return other is GUIDEInputMouseAxis2D;
	}

	public override string ToString() => "(GUIDEInputMouseAxis2D)";

	public override string _EditorName() => "Mouse Axis 2D";

	public override string _EditorDescription() => "Relative mouse movement on 2 axes.";

	public override GUIDEAction.GUIDEActionValueType _NativeValueType() => GUIDEAction.GUIDEActionValueType.Axis2D;

	public override DeviceType _DeviceType() => DeviceType.Mouse;
}
